x/subscription/keeper/subscription.go

```diff
@@ -76,63 +76,16 @@ func (k Keeper) CreateSubscription(
+		// If the plan index is different - upgrade if the price is higher
+		// If the plan index is the same but the plan block is different - treat as advanced purchase
 		if plan.Index != sub.PlanIndex {
-			currentPlan, err := k.GetPlanFromSubscription(ctx, consumer, block)
-			if err != nil {
-				return utils.LavaFormatError("failed to find plan for current subscription", err,
-					utils.LogAttr("consumer", consumer),
-					utils.LogAttr("block", block),
-				)
-			}
-			if plan.Price.Amount.LT(currentPlan.Price.Amount) {
-				return utils.LavaFormatError("New plan's price must be higher than the old plan", nil,
-					utils.LogAttr("consumer", consumer),
-					utils.LogAttr("currentPlan", currentPlan),
-					utils.LogAttr("newPlan", plan),
-					utils.LogAttr("block", block),
-				)
-			}
-			originalPlanIndex := sub.PlanIndex
-			originalPlanBlock := sub.PlanBlock
-			err = k.upgradeSubscriptionPlan(ctx, duration, &sub, &plan)
+			err := k.upgradeSubscriptionPlan(ctx, consumer, duration, &sub, &plan)
 			if err != nil {
 				return err
 			}
-
-			// Upgrade worked, now we can decrease the old plan's refcount
-			k.plansKeeper.PutPlan(ctx, originalPlanIndex, originalPlanBlock)
-
-			details := map[string]string{
-				"consumer": consumer,
-				"oldPlan":  sub.PlanIndex,
-				"newPlan":  plan.Index,
-			}
-			utils.LogLavaEvent(ctx, k.Logger(ctx), types.UpgradeSubscriptionEventName, details, "subscription upgraded")
 		} else if plan.Block != sub.PlanBlock {
-			// if the plan was changed since the subscription originally bought it, allow extension
-			// only if the updated plan's price is up to 5% more than the original price
-			subPlan, found := k.plansKeeper.FindPlan(ctx, plan.Index, sub.PlanBlock)
-			if !found {
-				return utils.LavaFormatError("cannot extend subscription", fmt.Errorf("critical: cannot find subscription's plan"),
-					utils.Attribute{Key: "plan_index", Value: plan.Index},
-					utils.Attribute{Key: "block", Value: strconv.FormatUint(sub.PlanBlock, 10)})
-			}
-			if plan.Price.Amount.GT(subPlan.Price.Amount.MulRaw(105).QuoRaw(100)) {
-				// current_plan_price > 1.05 * original_plan_price
-				return utils.LavaFormatError("cannot auto-extend subscription", fmt.Errorf("subscription's original plan price is lower by more than 5 percent than current plan"),
-					utils.Attribute{Key: "sub", Value: sub.Consumer},
-					utils.Attribute{Key: "plan", Value: plan.Index},
-					utils.Attribute{Key: "original_plan_block", Value: strconv.FormatUint(subPlan.Block, 10)},
-					utils.Attribute{Key: "original_plan_price", Value: subPlan.Price.String()},
-					utils.Attribute{Key: "current_plan_block", Value: strconv.FormatUint(plan.Block, 10)},
-					utils.Attribute{Key: "current_plan_block", Value: plan.Price.String()},
-				)
-			}
-
-			// update sub plan
-			sub.PlanBlock = plan.Block
+			return k.CreateFutureSubscription(ctx, creator, consumer, planIndex, duration)
```

Comment:
we have this flow already in the advance buy, if the user want to use advance buy let him use the flag.
instead return error here

Reply:
this flow lets the user use advance buy without using the flag  --advance-purchase
if the user wants to but something in advance let him do it with the flag so it will be explicit

Reply:
Fixed in 7fd7523aedabccabf7f8c7e5c34d25a7426e1e58